### ISSUE TITLE
Nano re-pairing is not possible and add a Quit Solaar button to also close tray icon

### ISSUE THREAD

**Issue by KenH2000** [pwr-Solaar/Solaar#662]
**The following code added to window.py will correct the issues**
1. If you only want to make a quick change to a device and not use the tray icon, you have to close the window, and then also close tray icon to exit the program.  A simple fix, is to add a 'Quit Solaar' button to the window.py source.  Edit ~/.local/lib/python3.6/site-packages/solaar/ui/window.py, after line 312 add:
```
#existing code...
	bottom_buttons_box.add(about_button)
#new code...
	close_button = _new_button(_("Quit") + ' ' + NAME, icon_size=_SMALL_BUTTON_ICON_SIZE, clicked=_justclose)
	bottom_buttons_box.add(close_button)
```
and add the function (also in window.py), which will close both the window and tray icon:
```
#new code...
def _justclose(self):
	import sys
	sys.exit()
```
Optional:  if you want space between the About and Quit buttons, edit line 310 to add spacing:
```
#existing code(comment out)...
	#bottom_buttons_box = Gtk.ButtonBox(Gtk.Orientation.HORIZONTAL)
#new code to add button spacing...
	bottom_buttons_box = Gtk.ButtonBox(Gtk.Orientation.HORIZONTAL,spacing=6)
```
2. If you pair a Nano device to another receiver, you can not re-pair it back to the original Nano receiver because the 'Pair new device' button is disabled, and cannot be re-enabled by unpairing the original device (unpairing always fails on Nano device).  To resolve this issue, force the button visible if the receiver name detected is a 'Nano' device.  Edit the window.py source, and edit the function, def _update_receiver_panel(...), after line 598 add the following:
```
#existing code...
		may_pair &= len(paired_devices) < receiver.max_devices
#new code...
		if "Nano" in receiver.name:
			may_pair=True        
#existing code....
	buttons._pair.set_sensitive(may_pair) 
	buttons._pair.set_visible(True)
```

**Comment by pfps** [comment]
@KenH2000 2. Yes, Solaar does not currently do a good job on nano re-pairing.  You should be able to re-pair with the CLI via `solaar pair` with doesn't have a check for empty pairing slots.  (If you have more than one receiver you have to specify which one.)  Expect to see an error message.  
I have changes to Solaar that affect re-pairing in my repository https://github.com/pfps/Solaar  which you can clone.  You should then checkout the c534 branch and run bin/solaar from the repository.  If your receiver is a c534 everything *should* work just fine.  If you have a different nano receiver you will need to tell solaar that it is a re-pairing receiver by fixing up its entry in base_usb.py.  You might be able to just use _nano_receiver_max2 but if your receiver has a different maximum you will have to create a new entry.  If you go this route please add your experiences to this issue.

**Comment by pfps** [comment]
@KenH2000 1.  I'll look into adding your changes to close Solaar directly.  I think hat this is a good idea.

**Comment by KenH2000** [comment]
Thank you for addressing my comments.  Solaar an awesome program, and equally awesome is that it is written in Python.  Before I found Solaar, I was using the Logitech program on a VM (it doesn't work in Wine) to configure Logitech receivers--which works--but I prefer to avoid windows whenever possible.  

I had already re-configured and tested my installation of Solaar, before I posted the comments.  I didn't want to post something that wouldn't work.  I like the _nano_receiver_max2 setting idea--I'll look into that, but for now, just force-enabling the button when 'Nano' is found in receiver.name, works great too.  The Nano receiver doesn't seem to care, and accepts the new pairing without issue.

**Comment by mickleby** [comment]
@KenH2000 @pfps in my (admittedly obstructed) view, "A simple fix, is to add a 'Quit Solaar' button to the window.py source," might better read, "A simple hack, is to add a 'Quit Solaar' button to the window.py source." I say this because, "want to make a quick change to a device and not use the tray icon, you have to close the window" looks like a serious flaw in itself. Thus we are chasing bugs with bugs, when we might well simply address the issue directly.

If the issue is this: Create a hack while we wait for the underlying resolution, I'm on board and understand the convenience of this suggested btnQuit hack. I understand there is a plan to revamp the architecture to leverage Dbus. This would resolve the underlying issue.

**Comment by KenH2000** [comment]
@mickleby I wouldn't call a close button a hack.  I probably would not have used the word 'hack' to describe my solution to a problem, but maybe 'patch' would have been a better word.  

Sometimes, these utilities that use the tray icon, come with a checkbox to enable or disable the tray feature.  But that would be a little more work to reprogram.

The Nano pairing button being greyed-out is probably a more serious problem that should be addressed with the GUI.  The GUI as-is (without my Nano hack--@mickleby, now this I would call a 'hack'), cannot be used to re-pair Nano devices, and you have to resort to the command line or a different program.  

In my opinion, this qualifies as a 'hack' because I don't see an elegant way around the problem with Nano devices.  In theory, you need to unpair the current device to get the count down to zero to enable the button.  But the current device will not unpair from a Nano.  But a new device will re-pair over the current device, as long as the button is enabled.

**Comment by mickleby** [comment]
@KenH2000 I'm concerned we're talking at cross purposes re the word "hack." I'll look again when I get home; perhaps I can be more clear.

I believe what I was trying to distinguish is
+ The need to run only a single instance of solaar
+ The utility of a btnQuit to facilitate closing solaar

The issue description leads me to determine "fix" will refer to resolving the first item, no longer requiring solaar to be closed.

Sorry for the confusion.

**Comment by mickleby** [comment]
@KenH2000 "a checkbox to enable or disable the tray feature. But that would be a little more work to reprogram" I think the skill and time are on par to the btnQuit. In your experience, how is the tray re-enabled? a CLI flag?

"cannot be used to re-pair Nano devices" Please bring me up to speed. Why can the device/receiver not pair from the GUI? Do you believe it is all Nano devices or your specific device?![pairing](http://2.bp.blogspot.com/--o4xQImAzlM/Ud_PGgg3yEI/AAAAAAAAPzs/k5RfVZ3d_RM/s320/solaar-logitech_2.png)

In your suggested code
```
#existing code...
		may_pair &= len(paired_devices) < receiver.max_devices
#new code...
		if "Nano" in receiver.name:
			may_pair=True    
```
I do not really follow the code together with your words before. You describe 2 receivers. This code is limited to a single receiver. The existing code is design to trap the situation where a user tries to pair more than the maximum number of devices a receiver may pair; your code says "pair 'em all" without limit. What does the GUI show when both receivers are connected? I'm guessing solaar expects only a single receiver, but I haven't tested this.

It seems to me you are describing a host of issues. I'm wondering if it would be helpful to redirect. Like, what is the scenario you would like solaar to enable?

Regarding only the quit button, I would expect a separate issue requesting a new feature is the proper avenue. @pfps has more experience with this community than I, so maybe the separate feature request is not needed.

(I'm off to read up on Nano receivers. cheers!)

**Comment by mickleby** [comment]
@KenH2000 BTW this sounds similar to your described 2 receiver situation https://github.com/pwr-Solaar/Solaar/issues/50#issuecomment-18140806

**Comment by KenH2000** [comment]
Not sure, but I think comment #50 is asking if it is possible to pair one device to more than one receiver simultaneously--which I'm pretty sure is not possible--for example, take your mouse to PC1, and then to PC2 and it would work in both, without having to pair it each time to PC1 and then to PC2.  

After reading your comments, I want to make sure we're on the same page.  When I'm talking about the 'Nano' device, I mean the Logitech Nano (capital 'N'), which is a one-device-only wireless receiver that supports pairing with other Logitech devices--but only one device at a time.  The word 'nano' can also refer to any unbranded wireless receiver.  The nano (little 'n') devices are not compatible with pairing software, or Logitech devices, and are permanently paired to the device they came with.

Logitech also makes a Unifying receiver.  The Unifying is like a Nano, but instead of only pairing with one device, it can support up to 6 Logitech devices at one time. 

Now that that is straight, the problem with the Nano, is that it will not unpair from its current device, like the Unifying will.  So there is no way to re-enable the Solaar pairing button, so you can pair a new device to the Nano.  But the Nano will re-pair to a new device (without first having to un-pair its current device), as long as the Solaar pairing button is enabled.

So my solution was to detect if 'Nano' was in the receiver's name, and force the button enabled.  

"What does the GUI show when both receivers are connected?  I'm guessing solaar expects only a single receiver, but I haven't tested this."

Solaar is able to show multiple receivers (make sure you have the latest version--maybe earlier versions didn't).  In the following scenario, we have one Nano and one Unifying connected.  The GUI shows both the Nano and the Unifying receivers in a drop down list.  When you choose a receiver, the may_pair code is invoked, and decides whether to enable or disable the pairing button.  When you choose a Unifying receiver, the may_pair code is invoked, and will display the pairing button as enabled if fewer than 6 devices are currently paired, and disabled if 6 devices are already paired.  This is fine for the Unifying, because it will allow you to unpair a device, the Solaar code will then re-enable the button, and you can continue to pair a new device.  

The problem occurs when you select the Nano from the list.  The original code will always show the pairing button disabled for a Nano, because the Nano always has a device paired, and cannot unpair it like the Unifying can.    My 'patch' forces the button enabled when a Nano device is chosen from the list.  The Nano WILL happily pair another device over the old one, but it won't unpair its current device, so with the code as written, you can never get the program to enable the button so you can pair a new device to your Nano.

I don't know how many kinds of Logitech Nano(s) exist.  I only have one to experiment with, and this is how mine works.

"In your experience, how is the tray re-enabled? a CLI flag?"

There are probably a hundred ways to do this...and I'm not sure any are as elegant as the quit button, but here's one idea.  Have a 'Preferences' menu item in the tray menu.  When you click it, have a pop-up with checkboxes, and you could have other stuff in there too, and save the settings in the config directory(that would be read and applied on program start), for example:

  **Solaar Preferences**
|&nbsp;X&nbsp;| enable tray icon
|&nbsp;&nbsp;&nbsp;&nbsp;| tray only (no pairing window on start) 
|&nbsp;X&nbsp;| always enable pairing button for Nano receivers
|&nbsp;&nbsp;&nbsp;&nbsp;| display a quit button in pairing window

**Comment by pfps** [comment]
@KenH2000 The c534 receiver is a Nano or Nano-like receiver but it can pair with a mouse and a keyboard (but only certain models of mice and keyboards).  It can't unpair, instead a new device replaces an old device of the same type.  So it looks very much like the receiver you have.  The c534 receiver also has only a limited number of repairings and once that limit is hit a c534 will no longer pair with new devices.  I have changes to Solaar in my github repository that handle the c534 and probably would easily handle your Nano reciever, including allowing pairing when no pairing slots are available.  
I've been holding off making a PR for these changes because they depend on a PR I already have set up to handle some minor aspects of receivers better.
So, I would suggest seeing how well these changes handle your receiver as I suggested above.

**Comment by KenH2000** [comment]
"The c534 receiver also has only a limited number of repairings..."
I was not aware of this.  If that's true for my Nano too, I can see why it is not a good idea to keep the pairing button enabled all the time.

**Comment by pfps** [comment]
@KenH2000 I don't know if it is true for your receiver.  I kind of suspect that it is true, however, and that the total number of pairings is 20.  What I have seen is that a register for the c534 is decremented on each pairing and when it hits a particular number (5, not zero!) the pairing lock cannot be opened any more.  Here is the code that I wrote to return the number of remaining pairings.  _R.receiver_connection is register 0x02.

	# how many pairings remain (None for unknown, -1 for unlimited)
	def remaining_pairings(self,cache=True):
		if self._remaining_pairings is None or not cache :
			ps = self.read_register(_R.receiver_connection)
			if ps is not None:
				ps = ord(ps[2:3])
				self._remaining_pairings = ps-5 if ps >= 5 else -1
		return self._remaining_pairings

This was all determined empirically.  I'm still waiting for information from Logitech on this register field.

**Comment by KenH2000** [comment]
Do you have that already written into one of your branches?  I'd like to see what number my Nano is at--that is, if you think it is safe to use.  Once the pairing lock is 'locked', does the Logitech tool also fail to re-pair?

**Comment by pfps** [comment]
@KenH2000 Yes, my c534 branch has all this in it.  You will need to set up your receiver as a re-pairing receiver in  base_usb.py as something like a _nano_receiver_max2 but without the max_devices bit.  Running this branch without doing any pairing should give you a number in the GUI if the field is used the same as it is for the c534.  I don't see how just running this branch would damage your receiver, as the register in question is already queried for other information, but, of course, YMMV.  One think you could do is see if the repairing information is shown and remember what it is.  Then you could go back to regular Solaar until you do re-pair.  Then run my branch and see if the number has changed.  There is a program (hidconsole) in the Solaar distribution that allows you to send commands (including register reads) directly to a receiver - you could also use that but you would have to figure out just what command to send - I've forgotten how this program works.
I did run the Logitech tool a while ago, but I don't remember exactly how it worked on c534 receivers and I've lost access to the couple of receivers that I ran out of re-pairing attempts.

**Comment by KenH2000** [comment]
I ran the C534 branch on my Nano.  I know you want me to add something to base_usb.py, but I don't understand what.   I added these lines, but I'm sure I'm missing something here:
```
_nano_receiver_max2 = lambda product_id: (0x046d, product_id, 2, _DRIVER)
# and further down
NANO_RECEIVER_MAX2        = _nano_receiver_max2(0xc52f)
# and further down in the ALL list
		NANO_RECEIVER_MAX2,
```
It comes up with the same ID as the NANO_RECEIVER_ADVANCED, and if I try to comment those lines out, and use the NANO_RECEIVER_MAX2 entries only, the receiver is not found.  Here is what the GUI reads (with and without the change in base_usb.py).  Nothing is shown about the max no. of pairings:
Path         : /dev/hidraw2
USB id       : 046d:c52f
Serial       : 58DB94BB
  Firmware   : 30.00.B0009
Notifications: wireless
               software present

**Comment by mickleby** [comment]
@KenH2000 I mention #50 specifically referring to the process described in this comment https://github.com/pwr-Solaar/Solaar/issues/50#issuecomment-18140806 Lekensteyn describes starting pair() on receiver1 and then sending unPair() to receiver2 then putting the keyboard into pairing mode. A 3 stage process, pair(1) > unPair(2) > keyboard (pairMode). I get the idea unPair(2) > pair(1) > keyboard(pairMode) is explicitly non-functional.

**Comment by pfps** [comment]
@KenH2000  The code for setting up the data structures for receivers in base_usb.py isn't the easiest to understand.

There is a simple change that should make all Nano receivers re-pair.  Just add
	're_pairs': True,
in base_usb.py after 
_nano_receiver = lambda product_id: {'vendor_id':0x046d,
(You may have to adjust indentation.)

That should be enough to make the Pair new device button active for any Nano receiver, and to do the right thing when a Nano receiver re-pairs.

On question: does the Solaar GUI window have the Unpair button active for your Nano receiver?
I didn't change the old code for determining whether a receiver can unpair and it might need some changes.

**Comment by KenH2000** [comment]
@pfps I think you are on step 2, while I'm struggling with step 1.  I was trying to find out how many re-pairs are left, using the c534 branch, but what code do I need to add to **first** make solaar recognize my Nano?  With no code changes yet, this branch doesn't even see my device.  The GUI doesn't show a device at all, and solaar show says:
```
Exception AttributeError: "'Receiver' object has no attribute '_devices'" in <object repr() failed> ignored
solaar: error: Logitech receiver not found
```
Or am I still on the wrong branch?  Should I be using the c534-new?
Sorry for all the confusion on my part  :-)

**Comment by pfps** [comment]
@KenH2000 c534-new isn't new.  I should have deleted it a while ago.  The right branch is c534.  If the  Solaar from this repository is recognizing your receiver then my branch c534 should also recognize it.

Maybe, just maybe, you are experiencing the python2/python3 problem.  As python2 is officially dead, support for python2 has just been dropped from Solaar but the dropping wasn't clean.  You might have to change the first line of bin/solaar to end with python3 instead of python.

**Comment by KenH2000** [comment]
@ pfps Good call.  The python3 change fixed it.  So here's what we have running bin/solaar show.  What do you make of the ' Has 34 successful pairing(s) remaining.' line?  I'm not sure I want to try pairing again, to see if it goes down to 33.
```
Nano Receiver
  Device path  : /dev/hidraw2
  USB id       : 046d:c52f
  Serial       : 58DB94BB
    Firmware   : 30.00.B0009
  Has 1 paired device(s) out of a maximum of 1.
  Has 34 successful pairing(s) remaining.
  Notifications: wireless, software present (0x000900)
```

**Comment by pfps** [comment]
@KenH2000 Do you see the
Exception AttributeError: "'Receiver' object has no attribute '_devices'" in <object repr() failed> ignored
solaar: error: Logitech receiver not found
when running regular Solaar?
Also please attach output of
solaar -dd show
from c534 branch.

**Comment by pfps** [comment]
@KenH2000 Please ignore previous comment.

**Comment by mickleby** [comment]
@pfps "Maybe, just maybe, you are experiencing the python2/python3 problem." Are there plans to cutoff python2?

**Comment by pfps** [comment]
@mickleby Yes, already done (as of a little while ago, more or less, as of today, much more than less).  PR #664

**Comment by mickleby** [comment]
@pfps again w my architect thing. My preference is to go back to the previous PR and update. This facilitates later review, whereas this way we have 2 separate PRs both claiming to achieve the same thing -- and neither actually achieving said without the other. How can some future reader decipher such?.. But I did locate #664 after I asked here. Cheers!

**Comment by pfps** [comment]
@mickleby Yeah, probably would have been better, but I don't know whether the commit log for Solaar can be fixed up this way.  I was going to edit a message or two to make things a bit clearer.

**Comment by mickleby** [comment]
@pfps "whether the commit log... can be fixed up this way" I suggest making a note in #664 that says "this PR intends to continue/resole" and link the other PR. Such notes then appear in the discussion being mentioned, and some later fellow reading will see to read there also

**Comment by pfps** [comment]
@KenH2000 The output you see looks reasonable.  My guess is that your receiver started out with about 40 pairings and you used up about 6 of them.  The c534 receivers appear to start out with 20 pairings (and I used up all of them on a couple of the receivers - there were quite a few sets of receiver/keyboard/mouse lying around at the lab I used to be in).  It would be very nice to see if this number goes down when you successfully pair with a new device.  What I see is that just opening the pairing lock doesn't make the number go down, it is pairing with a different device.
Another thing to check is what shows up in the GUI.  The Pair new device button should be active.   The Unpair button should not be active - but I don't trust the code that determines this so please tell me whether this button is active or not.  Thanks.

**Comment by KenH2000** [comment]
@pfps "The Pair new device button should be active. The Unpair button should not be active "
Just the opposite.  The **Unpair** button is active, and the **Pair new device** button is greyed out.  Maybe that's what you meant?  Because I would expect to be able to unpair one device (even though it won't work on a Nano) so I could then get to **zero** devices to enable the button, and then be able to pair a new device.  I wish it worked that way, but it doesn't on my Nano.  

I like the way the Unifying works a lot better--no pairing limits and it handles more than enough devices.  Well, at least you can buy one of Logitech's cheapest mice that includes a Unifying receiver for ~$10-$15, and forget about the Nano :-)

"... opening the pairing lock doesn't make the number go down, it is pairing with a different device."

So, do you think if I were to pair the original mouse again to its Nano receiver, the remaining_pairings would stay the same?  In other words, is it smart enough to know not to use a new slot, if a device ID is already in a pairing slot?  Or does it blindly keep adding device IDs (even if they are all the same IDs) until the slots are all used up.  I hope that question makes sense.

**Comment by pfps** [comment]
@KenH2000 I just realized I didn't have a branch that has all the changes needed to handle re-pairing for the c537 and also handle your gaming mouse.  I've pushed up a new branch - nano-c537
This branch is only for checking that the unpairing and pairing stuff works for your receiver so you will probably have to have your mouse off.  With this branch the Pair new ... button for your receiver should be active and the Unpair button for your mouse should be inactive.  If the former is not the case I have a bug somewhere.  If the latter is not the case then there is a bug in some old Solaar code.  Let me know what you see. 
I think that to get your receiver and mouse into Solaar there will have to be a demonstration that pairing works correctly on it.
PS:  I've force-pushed some stuff to my Solaar repository.  This  might interfere with simply updating your clone.  (I'm not sure how git works in this case.)

**Comment by pfps** [comment]
@KenH2000 I don't have complete documentation on pairing for nano receivers.  I have seen on my c534 receivers that the only thing that decrements the remaining-pairings counter is successful changing from one device to another or adding a device to an empty slot.  (The c534 has two slots and some of them come paired with only one device.)  I think that switching between two mice always decrements the counter.   I wouldn't count on switching back to the *original* mouse not decrementing the counter or even being allowed when the counter is a 0.
The way that these receivers work, as far as I can tell, is that pairing generally puts the new device in the same slot as the old one.   (Most nano receivers only have one slot anyway.)

**Comment by KenH2000** [comment]
@pfps "Let me know what you see."
Both buttons are active.  That's OK.  I have that M325 mouse paired to its original Nano, and am going to leave it alone now.  I don't want to lose any more pairings on that Nano/mouse combo.  I have another mouse that came with a Unifying, that I can use with other receivers.

**Comment by KenH2000** [comment]
@pfps I had a revelation last night.  I would alter the _pair_new_device function in window.py, to warn about the pairing limit for Nano receivers, and give the user one last chance to cancel before pairing a Nano.  You could make the message more informative by showing the number of remaining pairings in the dialog, but I didn't do that in this example:
```
	def _pair_new_device(trigger):
		assert _find_selected_device_id() is not None
		receiver = _find_selected_device()
		assert receiver is not None
		assert bool(receiver)
		assert receiver.kind is None
#new code
		if 'Nano' in receiver.name:
			q = Question("Nano receivers have a limited\nnumber of pairings.\n")
			if q.returncode < 0: #if user cancels or closes dialog, pairing action is skipped
				return
#end new code       
		_action.pair(_window, receiver)
```
Then add the dialog class for above code somewhere.  I put it at the end of window.py, just to see if it would work.  I didn't go through all your code to see if you already have a dialog class somewhere else you could use.  If so, you could use that instead of adding this code.
```
#new code
class Question():

	def __init__(self,mycontent):
		self.message_dialog = Gtk.Dialog(
			"Warning",
			None,
			Gtk.DialogFlags.MODAL,
			)

		box = self.message_dialog.get_content_area()
		label = Gtk.Label(mycontent)
		box.add(label) 
		self.message_dialog.add_buttons("Cancel",-1,"Continue",1)
		self.message_dialog.show_all()
		response = self.message_dialog.run()
		self.returncode = response
		self.message_dialog.destroy()
```

**Comment by pfps** [comment]
@KenH2000 Let me think about this for a while.

**Comment by pfps** [comment]
@KenH2000 The merged PR #674 adds the above warning.  PR #678 does a better determination of when to show the warning.

**Comment by pfps** [comment]
PR #679 and #675 should cover all the issues here
